### ISSUE TITLE
Prisma creates implicit `ORDER BY` in query

### ISSUE THREAD

**Issue by IamR3m** [prisma/prisma#21032]
### Bug description

When I create a query with some filter and select fields but without orderBy (even if it explicitly set to `undefined`), then prisma forcibly adds `ORDER BY id ASC` in resulting query to DB, which slows down a query in comparison to `$rawQuery` without `ORDER BY`

### How to reproduce

1. Create prismaClient with logging queries. 
2. Create 2 queries. One with prisma syntax, another one with rawQuery.
3. Add logging time spent to execute queries.
4. Run your script.
5. See logs.

### Prisma information

```ts
const prisma = new PrismaClient({log: ["query"]});

// warmup connection to DB
await prisma.posts.count({where: {title: {contains: "prisma"}}});

// Prisma syntax query
console.time("Prisma");
const prismaRes = await prisma.posts.findMany({
  select: {
    title: true,
    views: true
  },
  where: {
    title: {contains: "prisma"}
  },
  take: 10
});
console.timeEnd("Prisma");

// Raw query
console.time("Raw");
const query = Prisma.sql([`
  SELECT title, views
  FROM posts
  WHERE title LIKE "%prisma%"
  LIMIT 10
`]);
const rawRes = await prisma.$queryRaw(query);
console.timeEnd("Raw");
```

### Environment & setup

- OS: Ubuntu 22.04.3 LTS
- Database: MySQL 8.0.26
- Node.js version: 18.17.1

### Prisma Version

```
5.2.0
```

**Comment by IamR3m** [comment]
@janpio No. Without `take: 10` there's no `ORDER BY` anymore but it's a bad solution because our tables have 1M+ records and we have to use pagination.

**Comment by janpio** [comment]
I didn't want to suggest that this is the solution, just get context.

The `ORDER BY` makes sure that you always get the same 10 results. Otherwise, this would not be guaranteed.
Optimally you should create the appropriate indexes that make this query equally fast as the one without the guarantee.

**Comment by IamR3m** [comment]
I already added indexes to all fields we can use for filtering. How can I disable this implicit behaviour with `ORDER BY`?

**Comment by janpio** [comment]
Right now you can't, this is not an option in Prisma as we need to guarantee the order as soon as `take` or `skip` come into play - otherwise running the same query twice could change the dataset although no actual change happened in the expected data. There are not guarantees from the database itself.

Can you provide some more information on your Prisma schema and database SQL? I am confused why this could have a performance impact and would like to investigate to understand. Thanks.

**Comment by tupe12334** [comment]
Hi @janpio According to my understanding there is a product requirement that says "For every identical query the data that returns needs to be the same (if the data in the DB didn't change)"
Can you point me to the spec test that checks that? and if not, there is an open issue about that, I will be more than happy to work on that.

**Comment by janpio** [comment]
I don't know if there is such a spec test. The above is based on conversations with some engineers, and a experimental PR here that shows that this (with the current tests) only causes issues for SQL Server: https://github.com/prisma/prisma-engines/pull/4233 But we still think that the ordering is relevant to achieve the intended functionality of stable sorting on paginated queries.

**Comment by IamR3m** [comment]
Please, let me know if https://github.com/prisma/prisma-engines/pull/4233 will be released. But it'll be better if this will be switched by some option in `prisma.schema` or `PrismaClientOptions`.

**Comment by janpio** [comment]
It will not, as it makes some tests fail and changes how Prisma behaves. The PR was just to test how it behaves exactly.

**Comment by janpio** [comment]
Internal note: Confirm ordering behavior with and without `take` in query via SQL query logging.

**Comment by askkaz** [comment]
We ran into this at some scale with the count function and we're having to switch over to raw.
Here is a rough schema to help you visualize how performance is impacted:
```
model Posts {
  id  String  @id @default(cuid())
  authorId  String
  author  Author @relation(fields: [authorId], references: [id], onDelete: Cascade)
  @@index([authorId])
}
```
Our use case - we wanted to quickly determine if a user had at least 3,000 posts. We used a query like:
```
await prisma.posts.count({
  where: {
    authorId: "some_id"
  },
  take: 3_000
});
```
The query planner decided to use the Post id index first due to the order by and then filter down to the team id. Without the order by in our case, the query planner used the authorId index.

I also can't think of a reason why "order by" would be needed in a count function - I'm assuming its due to re-using some code for aggregate functions in prisma.

**Comment by janpio** [comment]
Good observation @askkaz. Can you maybe create a separate issue for the case of the `count()` query where an order is not even surfaced in the returned data? Thank you.

**Comment by CaioSchmitt** [comment]
we have the same problem with queries that use take, our table is very big and it is causing slow query performance because prisma uses the order by automatic.
 have you solved this problem? @IamR3m

**Comment by IamR3m** [comment]
@CaioSchmitt nope, sorry. still waiting...

**Comment by mateuspiresl** [comment]
I understand that the Prisma team's intention is to keep it consistent, but I'm trying to come up with another perspective here.

> does this really make the query slower when correct indexes are in place?

Even if a correct index can prevent performance issues in this specific query, it may add extra load to the table writes. Also, there will be more indexes to maintain. Large applications with tightly optimized indexes are more likely to suffer from that.

That is an extra index (or larger index) added to meet a requirement that didn't come from the product itself but from a leaked implementation detail of the library it uses.

Using ORDER BY to produce consistent query results is a well-known approach. IMO, the developer must be responsible for implementing it when needed. I think that would make an ideal optional behavior in Prisma.

**Comment by bartoszhernas** [comment]
Hi, I just stumbled upon this and it breaks my code. My table has 100m records. I need to fetch records with empty column to process them.

Easy peasy, normal SQL runs in 30ms
```sql
SELECT id from "OriginalTrack" where "trackId" is null LIMIT 30;
```

Prisma takes  ~5minutes per query which is insane. I have 73 million records to process. Documentation about ordering mentions nothing about forced "ORDER BY" clause, thank god I stumbled upon here. 

Nowhere in the documentation is there a requirement that this will return same 10 records, but I understand the point. Then allow me to force the query planner to ignore this shadow-hidden requirement and force it to drop `ORDER BY`. 

Good thing would also be to have this in documentation. I saw the ORDER BY is added because in processes list it was plain as day that there is a query stuck. **I've read all the page about sorting and couldn't find anything about having this added by default. ChatGPT knew fortunately to link to this page**

I guess for now I need to switch back to raw query sadly

**Comment by Thomasvdam** [comment]
Just chiming in to say I ran into the same issue. A count query which should be fast took forever and after some digging I noticed that for some reason Prisma adds `ORDER BY` to the count query. I think documenting this behaviour is already helpful, and even better would be a built-in escape hatch. For now I'll also switch to a raw query :(

**Comment by zyf0330** [comment]
This is a stupid implict behavior, which is a big performance problem when transfer huge table data by a field in batch.
And I cannot find a way to disable it, otherwise I must write raw sql.
